File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RewriteRuleCompositionOptions.java
package com.foundgine.core.semantic.planning;

/**
 * Safety, determinism and selection limits for composing semantic rewrite
 * rules.
 */
public record RewriteRuleCompositionOptions(int maxRuleApplications, int maxPlanVisits,
		RuleSelectionPolicy selectionPolicy, IProviderCostEstimator providerCostEstimator,
		ProviderCostSelectionPolicy providerCostSelectionPolicy) {
	public RewriteRuleCompositionOptions() {
		this(32, 64, null, null, null);
	}

	public RewriteRuleCompositionOptions(int maxRuleApplications, int maxPlanVisits) {
		this(maxRuleApplications, maxPlanVisits, null, null, null);
	}

	public RewriteRuleCompositionOptions validate() {
		if (maxRuleApplications < 1)
			throw new IllegalArgumentException("maxRuleApplications must be positive.");
		if (maxPlanVisits < 1)
			throw new IllegalArgumentException("maxPlanVisits must be positive.");
		(selectionPolicy == null ? new RuleSelectionPolicy() : selectionPolicy).validate();
		(providerCostSelectionPolicy == null ? new ProviderCostSelectionPolicy() : providerCostSelectionPolicy)
				.validate();
		if (providerCostEstimator != null
				&& (providerCostEstimator.provider() == null || providerCostEstimator.provider().isBlank()))
			throw new IllegalArgumentException("Provider cost estimator must identify its provider.");
		return this;
	}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/RewriteRuleCompositionOptions.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Safety, determinism and selection limits for composing semantic rewrite rules.</summary>
public sealed record RewriteRuleCompositionOptions(
    int MaxRuleApplications = 32,
    int MaxPlanVisits = 64,
    RuleSelectionPolicy? SelectionPolicy = null,
    IProviderCostEstimator? ProviderCostEstimator = null,
    ProviderCostSelectionPolicy? ProviderCostSelectionPolicy = null)
{
    public RewriteRuleCompositionOptions Validate()
    {
        if (MaxRuleApplications < 1) throw new ArgumentOutOfRangeException(nameof(MaxRuleApplications));
        if (MaxPlanVisits < 1) throw new ArgumentOutOfRangeException(nameof(MaxPlanVisits));
        (SelectionPolicy ?? new RuleSelectionPolicy()).Validate();
        (ProviderCostSelectionPolicy ?? new ProviderCostSelectionPolicy()).Validate();
        if (ProviderCostEstimator is not null && string.IsNullOrWhiteSpace(ProviderCostEstimator.Provider))
            throw new ArgumentException("Provider cost estimator must identify its provider.",
                nameof(ProviderCostEstimator));
        return this;
    }
}